File: src/main/java/at/ac/uibk/dps/cirrina/execution/command/ActionMatchCommand.java
package at.ac.uibk.dps.cirrina.execution.command;

import at.ac.uibk.dps.cirrina.execution.object.action.MatchAction;
import com.google.common.flogger.FluentLogger;
import java.util.ArrayList;
import java.util.List;

public final class ActionMatchCommand extends ActionCommand {

  private static final FluentLogger logger = FluentLogger.forEnclosingClass();

  private final MatchAction matchAction;

  ActionMatchCommand(ExecutionContext executionContext, MatchAction matchAction) {
    super(executionContext);
    this.matchAction = matchAction;
  }

  @Override
  public List<ActionCommand> execute() throws UnsupportedOperationException {
    final var commands = new ArrayList<ActionCommand>();

    try {
      final var extent = executionContext.scope().getExtent();
      final var conditionValue = matchAction.getValue().execute(extent);

      final var commandFactory = new CommandFactory(executionContext);

      // Find matching conditions and append the commands to the set of new commands
      for (var entry : matchAction.getCase().entrySet()) {
        final var caseValue = entry.getKey().execute(extent);
        final var caseAction = entry.getValue();

        // In case the case condition matches, add the case action
        if (conditionValue == caseValue) {
          final var command = commandFactory.createActionCommand(caseAction);

          commands.add(command);
        }
      }
    } catch (UnsupportedOperationException e) {
      logger.atWarning().withCause(e).log("Could not execute match action");
    }

    return commands;
  }
}

File: src/main/java/at/ac/uibk/dps/cirrina/execution/command/ActionMatchCommand.kt
package at.ac.uibk.dps.cirrina.execution.command

import at.ac.uibk.dps.cirrina.execution.`object`.action.MatchAction
import io.micrometer.core.instrument.MeterRegistry

/**
 * A command that evaluates a match expression and executes the corresponding action(s) within the
 * provided [executionContext].
 *
 * It evaluates a central value and compares it against a set of case expressions. If matches are
 * found, those actions are converted into commands; otherwise, the default action is used.
 *
 * @property matchAction the definition containing the value to match and the cases.
 * @property executionContext the context providing scope and command creation capabilities.
 * @property meterRegistry the registry used for collecting metrics.
 */
class ActionMatchCommand
internal constructor(
  private val matchAction: MatchAction,
  executionContext: ExecutionContext,
  meterRegistry: MeterRegistry,
) : ActionCommand(executionContext, meterRegistry) {

  /**
   * Executes the match logic.
   *
   * @return a list of [ActionCommand]s to be executed.
   * @throws Exception if the command execution fails due to an internal error.
   */
  override fun execute(): List<ActionCommand> =
    matchAction.cases.entries
      .filter { (expression, _) ->
        expression.execute(executionContext.scope.extent) ==
          matchAction.value.execute(executionContext.scope.extent)
      }
      .map { it.value }
      .ifEmpty { listOfNotNull(matchAction.default) }
      .map { action -> CommandFactory(executionContext, meterRegistry).createActionCommand(action) }
}
